tinygrad/tensor.py

```diff
@@ -1,34 +1,67 @@
 # inspired by https://github.com/karpathy/micrograd/blob/master/micrograd/engine.py
 from __future__ import annotations
-import math, functools, itertools, operator, time
+import time
+from functools import partialmethod, reduce
+from itertools import accumulate, filterfalse
+import operator
 import numpy as np
-from typing import List, Tuple, Callable, Optional, ClassVar, Type, Union, Sequence
-from tinygrad.helpers import prod, argfix, make_pair, getenv, IMAGE, DEBUG, flatten, DType, dtypes, ImageDType
+from typing import List, Tuple, Callable, Optional, ClassVar, Type, Union, Sequence, cast
+from tinygrad.helpers import argfix, make_pair, getenv, IMAGE, DEBUG, flatten, DType, dtypes
+from math import ceil, pi, prod, sqrt
 from tinygrad.lazy import Device, LazyBuffer
 from tinygrad.ops import LoadOps
 
 # An instantiation of the Function is the Context
 class Function:
-  def __init__(self, device:str, *tensors:Tensor):
-    self.device, self.parents = device, tensors
-    self.needs_input_grad = [t.requires_grad for t in self.parents]
-    self.requires_grad = True if any(self.needs_input_grad) else (None if any(x is None for x in self.needs_input_grad) else False)
+  __slots__ = "device", "parents", "needs_input_grad", "requires_grad"
+  device: str
+  parents: Tuple[Tensor, ...]
+  needs_input_grad: Tuple[Optional[bool], ...]
+  requires_grad: Optional[bool]
 
+  @classmethod
+  def apply(fxn:Type[Function], *args, **kwargs) -> Tensor: raise NotImplementedError(f"apply not implemented for {type(fxn)}")
+  def forward(self, *args, **kwargs): raise NotImplementedError(f"forward not implemented for {type(self)}")
+  def backward(self, *args, **kwargs): raise RuntimeError(f"backward not implemented for {type(self)}")
+
+class UnaryFunction(Function):
+  def __init__(self, device:str, a:Tensor):
+    self.device, self.parents = device, (a,)
+    self.needs_input_grad = (a.requires_grad,)
+    self.requires_grad = a.requires_grad
+
+  @classmethod
+  def apply(fxn:Type[UnaryFunction], a:Tensor, **kwargs) -> Tensor:
+    ctx = fxn(a.device, a)
+    ret = Tensor(ctx.forward(a.lazydata, **kwargs), device=ctx.device, requires_grad=ctx.requires_grad)
+    if ctx.requires_grad and not Tensor.no_grad: ret._ctx = ctx    # used by autograd engine
+    return ret
+  
   def forward(self, *args, **kwargs): raise NotImplementedError(f"forward not implemented for {type(self)}")
   def backward(self, *args, **kwargs): raise RuntimeError(f"backward not implemented for {type(self)}")
 
+class BinaryFunction(Function):
+  def __init__(self, device:str, a:Tensor, b:Tensor):
+    self.device, self.parents = device, (a, b)
+    self.needs_input_grad = (a.requires_grad, b.requires_grad)
+    self.requires_grad = True if any(self.needs_input_grad) else (None if any([x is None for x in self.needs_input_grad]) else False)
```

Comment:
What about `self.requires_grad = True if True in self.needs_input_grad else None if None in self.needs_input_grad else False`? Should be faster.